module/zfs/zap_micro.c

```diff
@@ -363,6 +363,43 @@ mze_find_unused_cd(zap_t *zap, uint64_t hash)
 	return (cd);
 }
 
+/*
+ * Each mzap entry requires at max : 4 chunks
+ * 3 chunks for names + 1 chunk for value.
+ */
+#define	MZAP_ENT_CHUNKS	(1 + ZAP_LEAF_ARRAY_NCHUNKS(MZAP_NAME_LEN) + \
+	ZAP_LEAF_ARRAY_NCHUNKS(sizeof (uint64_t)))
+
+/*
+ * Check if the current entry keeps the colliding entries under the fatzap leaf
+ * size.
+ *
+ * This is important to ensure that the microzap is upgradable to fatzap.
```

Comment:
[nit] I'm not sure this needs to be its own paragraph, you could include it after the sentence above.

Reply:
Deleting that line as it is no longer correct. As per Matt's suggestion, we now trigger an upgrade to fatzap when the number of chunks exceed the limit.

Without this change we were failing the zap_add() for a smaller number if it was a microzap.
With upgrade, we would first upgrade to fatzap and that would allow one to add more entries with colliding hash. And eventually zap_add() to fatzap would fail when the colliding entry can longer fit in the leaf-block.